File: spring-boot/src/main/java/com/erdaldalkiran/cart/domain/cart/LineItem.java
package com.erdaldalkiran.cart.domain.cart;

import com.erdaldalkiran.cart.domain.campaign.Campaign;
import com.erdaldalkiran.cart.domain.product.Product;
import lombok.Getter;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;

@Getter
public class LineItem {
    private final Product product;
    private final int quantity;
    private Campaign appliedCampaign;
    private BigDecimal couponDiscount;

    public LineItem(Product product, int quantity) {
        this.product = product;
        this.quantity = quantity;
    }

    public BigDecimal getCampaignDiscount() {
        if (appliedCampaign == null) {
            return BigDecimal.ZERO;
        }

        return appliedCampaign.calculateDiscountAmountFor(this)
                .orElse(BigDecimal.ZERO);
    }

    public BigDecimal getTotalDiscount() {
        return getCampaignDiscount().add(couponDiscount).round(new MathContext(2, RoundingMode.HALF_EVEN));
    }

    public BigDecimal getTotalAmount() {
        return product.getPrice().multiply(BigDecimal.valueOf(quantity)).round(new MathContext(2, RoundingMode.HALF_EVEN));
    }

    public BigDecimal getTotalAmountAfterCampaignDiscount() {
        return getTotalAmount().subtract(getCampaignDiscount()).round(new MathContext(2, RoundingMode.HALF_EVEN));
    }

    public BigDecimal getTotalAmountAfterDiscounts() {
        return getTotalAmount().subtract(getTotalDiscount()).round(new MathContext(2, RoundingMode.HALF_EVEN));
    }

    public void clearCampaign() {
        appliedCampaign = null;
    }

    public void applyCampaign(Campaign campaign) {
        var isApplicable = campaign.isApplicableTo(this);
        if (!isApplicable) return;
        appliedCampaign = campaign;
    }

    public void clearCouponDiscount() {
        couponDiscount = BigDecimal.ZERO;
    }

    public void setCouponDiscount(BigDecimal amount) {
        couponDiscount = amount.round(new MathContext(2, RoundingMode.HALF_EVEN));
    }
}

File: dotnet/src/ShoppingCart.Business/Cart/LineItem.cs
﻿using System;

namespace ShoppingCart.Business.Cart
{
    public class LineItem
    {
        public Product.Product Product { get; }
        public int Quantity { get; }
        public Campaign.Campaign AppliedCampaign { get; private set; }
        public decimal CampaignDiscount => AppliedCampaign?.CalculateDiscountAmountFor(this) ?? 0m;
        public decimal CouponDiscount { get; private set; }
        public decimal TotalDiscount => Math.Round(CampaignDiscount + CouponDiscount, 2);
        public decimal TotalAmount => Math.Round(Product.Price * Quantity, 2);
        public decimal TotalAmountAfterCampaignDiscount => Math.Round(TotalAmount - CampaignDiscount, 2);
        public decimal TotalAmountAfterDiscounts => Math.Round(TotalAmount - TotalDiscount, 2);

        public LineItem(Product.Product product, int quantity)
        {
            Product = product;
            Quantity = quantity;
        }

        public void ClearCampaign()
        {
            AppliedCampaign = null;
        }

        public void ApplyCampaign(Campaign.Campaign campaign)
        {
            var isApplicable = campaign.IsApplicableTo(this);
            if (!isApplicable) return;
            AppliedCampaign = campaign;
        }

        public void ClearCouponDiscount()
        {
            CouponDiscount = 0m;
        }

        public void SetCouponDiscount(decimal amount)
        {
            CouponDiscount = Math.Round(amount, 2);
        }
    }
}